Fix #2203 (combine with list of characters with NA)

zeehio opened the pull request:
This rather small pull request:
- Adds a test case showing issue #2203 
- Provides a fix for #2203. This fix increases the compatibility of the character collector to tolerate logical `NA` and converting them to `NA_character_`.
- Adds the corresponding bullet to NEWS file

Any feedback will be very much appreciated

krlmlr commented:
LGTM. Fixes #2203.

hadley commented:
I wonder if it would be better to be stricter and only check for logicals of length 1?

Do we need to apply the same principle for other atomic vectors, or are they already covered? Either way, it's probably worth pulling `LGLSXP == TYPEOF(x) && all_na(x)` out into an informatively named function.

zeehio commented:
Logicals with all `NA` of length > 1 also are meaningful to me:
```r
dplyr::combine(list("a", "b", "c", c(NA, NA), "e")) 
```
Should work like:
```r
dplyr::combine(list("a", "b", "c", c(NA_character_, NA_character_), "e"))
```
Returning
```r
c("a", "b", "c", NA, NA, "e")
```

I will write tests for the other collecters to check if this happens to them too and prevent regressions. I will pull `LGLSXP == TYPEOF(x) && all_na(x)` into its own function named `all_logical_na` so I can use it in other collecters if required.

krlmlr commented:
```r
> combine(list(1:3, NA))
[1]  1  2  3 NA
> combine(list(1.5:3, NA))
[1] 1.5 2.5  NA
> combine(list(complex(real = 1.5:3), NA))
Error in eval(substitute(expr), envir, enclos) : 
  Can not automatically convert from complex to logical.
> combine(list(factor(1.5:3), NA))
Error in eval(substitute(expr), envir, enclos) : 
  Can not automatically convert from factor to logical.
```

zeehio commented:
Test and fix if necessary that combine handles missing values with:

 - [x] logical
 - [x] integer
 - [ ] factor
 - [x] real
 - [x] character
 - [ ] POSIXct
 - [ ] Date
 - [ ] complex

```
  OK: 1609 SKIPPED: 3 FAILED: 4
  1. Error: combine works with NA and factors (#2203) (@test-combine.R#49) 
  2. Error: combine works with NA and POSIXct (#2203) (@test-combine.R#86) 
  3. Error: combine works with NA and Date (#2203) (@test-combine.R#103) 
  4. Error: combine works with NA and complex (#2203) (@test-combine.R#131) 
```

hadley commented:
My main concern with checking the value for vectors > length 1 is performance since you're now scanning the value of every element. But I guess it can't actually cause a performance regression, since currently the code throws an error.

krlmlr commented on inst/include/dplyr/Collecter.h:
```diff
@@ -33,10 +37,16 @@ namespace dplyr {
-      STORAGE* source_ptr = Rcpp::internal::r_vector_start<RTYPE>(source);
-      for (int i=0; i<index.size(); i++) {
-        data[index[i]] = source_ptr[i];
+      if (all_logical_na(v)) {
```
I like the way Collecter<STRSXP>::collect() is split into functions with clear names.

krlmlr commented on inst/include/dplyr/Collecter.h:
```diff
@@ -295,19 +319,25 @@ namespace dplyr {
-      int* source_ptr = Rcpp::internal::r_vector_start<INTSXP>(source);
-      for (int i=0; i<index.size(); i++) {
-        if (source_ptr[i] == NA_INTEGER) {
+      if (Rf_inherits(v, "factor") && has_same_levels_as(v)) {
```
Same.

zeehio commented:
This still requires further work, some cases where there is an initial NA fail:

```r
combine(list(NA, "hello"))
# Error cannot convert from logical to character
``` 

Similar errors happen with factors and complex.

Solutions (I need to work on them):
- Promote logical collecter to character collecter if the logical collecter only has NA.
- Promote logical collecter to factor collecter if the logical collecter only has NA.
- Promote logical collecter to complex collecter. Would you like to convert FALSE to 0+0i and TRUE to 1+0i as well in this case?

krlmlr commented on inst/include/dplyr/Collecter.h:
```diff
@@ -9,6 +9,10 @@
 
 namespace dplyr {
 
+  static inline bool all_logical_na(SEXP x) {
```
Can we somehow reuse is_logical_all_na() below?

krlmlr commented:
I think the collecter should be strict and only mix logicals with other types (including numerics) if only NAs were seen.

I remember fixing very similar issues in DelayedProcessor.h.

zeehio commented:
@krlmlr,

- I hope I understood your comments in the code review. I tried to address both splitting the `collect()` function and the reuse of `is_logical_all_na()`. If I misunderstood you, please correct me again.
- If we change the collecter to be strict with respect to numbers then we break backwards compatibility. I will document this breaking change in the NEWS file.
- I still have some test cases to fix, but it's time for me to go to sleep. I will try to fix this tomorrow (in approx 20 hours).

Thanks for all your feedback, comments and suggestions

krlmlr commented:
@hadley: What should `combine(list(1, TRUE, FALSE, 2))` do -- raise an error or return `c(1, 1, 0, 2)` ?

hadley commented on tests/testthat/test-combine.R:
```diff
@@ -16,3 +16,181 @@ test_that("combine works with input that used to fail (#1780)", {
+
+test_that("combine works with NA and Date (#2203)", {
+  # NA first
+  expected_result <- as.Date(c(NA, "2010-01-01", "2010-01-02", NA, "2010-01-04"))
```
I think this would be a little easier to read if you initialised as `as.Date("2010-01-01") + c(NA, 0, 1, NA, 3)` etc

hadley commented on tests/testthat/test-combine.R:
```diff
@@ -16,3 +16,181 @@ test_that("combine works with input that used to fail (#1780)", {
+  expected_result <- as.Date(c(NA, "2010-01-01", "2010-01-02", NA, "2010-01-04"))
+  works1 <- combine(list(NA, as.Date("2010-01-01"), as.Date("2010-01-02"),
+                         as.Date(NA), as.Date("2010-01-04")))
+  expect_equal(works1, expected_result)
```
I think this could just be `expect_equal(combine(as.list(works1)), works1)`

hadley commented on tests/testthat/test-combine.R:
```diff
@@ -16,3 +16,181 @@ test_that("combine works with input that used to fail (#1780)", {
+
+  # NA length > 1
+  expected_result <- c(as.Date(c("2010-01-01", "2010-01-02", NA, NA, "2010-01-04")))
+  works1 <- combine(list(as.Date("2010-01-01"), as.Date("2010-01-02"),
```
Here you could maybe make the principle easier to understand by doing `split(works1, c(1, 2, 3, 3, 4))` or similar.

hadley commented on tests/testthat/test-combine.R:
```diff
@@ -16,3 +16,181 @@ test_that("combine works with input that used to fail (#1780)", {
+  expect_equal(works1, expected_result)
+
+  # NA length == 1
+  expected_result <- complex(real = c(1, 2, NA, 4),
```
`c(1, 2, NA, 4) + 1i` would be more compact

hadley commented:
@krlmlr For now I'd say it should be an error because we want to be strict about coercing types (integer -> double is one exception). We will need to write up the exact principles in vctrs - currently we have somewhat behaviour spread across multiple functions (i.e. https://github.com/hadley/vctrs/issues/7).

zeehio commented:
- I have improved the readability of the tests following @hadley's suggestions. 
- I have added a test case that loops through all pairs of atomic types and combines them, checking that the result or error matches the expected result. Following a "strict coercing types policy" I just allow coercing NA values to anything, integers to reals, and factors to characters.
- I have fixed all the test cases dealing with combine. However I now have a test case failing:

```r
# (@test-binds.R#123)
test_that("bind_rows promotes logical to integer", {
  df1 <- data_frame(a = FALSE)
  df2 <- data_frame(a = 1L)

  res <- bind_rows(df1, df2)
  expect_equal(res$a, c(0L, 1L))
})

```

So the question now is, do we remove the coercion from logical to integer in `bind_rows` or do we use different rules in each case?

In case you decide to remove the coercion I will replace the expect_equal in the test above to an expect_error.

Before merging, the NEWS entry will need to be updated. Probably tomorrow.

krlmlr commented on inst/include/dplyr/Collecter.h:
```diff
@@ -177,11 +210,13 @@ namespace dplyr {
       int RTYPE = TYPEOF(x);
-      return (INTSXP == RTYPE || RTYPE == LGLSXP) && !Rf_inherits(x, "factor");
+      return (INTSXP == RTYPE && !Rf_inherits(x, "factor") &&
+                !Rf_inherits(x, "POSIXct") && !Rf_inherits(x, "Date")) ||
```
This isn't completely failsafe against other classes such as `"difftime"`. I wonder if the additional checks here are worth the effort, or if we should just let it slide and wait for hadley/vctrs#7 (and potentially hadley/vctrs#27). @hadley?

hadley replied:
I think we should probably treat any vector with a class as incompatible by default, but I agree we don't need to do that in this PR.

zeehio replied:
I did not implement support for "difftime", but now it treats any integer or real vector with a class as incompatible (0b9ed6a257326930a3e052a29b3bdcc95e60e174) so if someone tries to collect an unsupported class (such as "difftime" or "dummyint" -in the test case-) the collecter returns an error. If you are not sure about this behaviour I can revert that and custom classes will be coerced (with potential loss of information).

zeehio replied:
Maybe raising a warning and then coercing the class is a better solution for now (breaks less code and allows users to adapt).

I could also allow coercion from logical to integer with a warning, so the breaking change is softer. The decision is up to you. :-)

krlmlr commented on NEWS.md:
```diff
@@ -63,6 +63,8 @@
 * `mutate_all()` etc now accept unnamed additional arguments.
 
+* `combine()` accepts `NA` values (#2203, @zeehio)
+
```
Breaking changes should be documented clearly.

zeehio replied:
Done in 9fb42da

zeehio commented on tests/testthat/test-combine.R:
```diff
@@ -16,3 +16,219 @@ test_that("combine works with input that used to fail (#1780)", {
+  # NA first
+  expected_result <- factor(c(NA, "a", "c", NA, "b"), levels = c("a","b","c"))
+  works1 <- combine(list(NA,
+                         factor("a", levels = c("a","b","c")),
```
Done eaa9b0d10463f6d885f760187f2094a7949e1d02

zeehio commented on tests/testthat/test-combine.R:
```diff
@@ -16,3 +16,219 @@ test_that("combine works with input that used to fail (#1780)", {
+  # Date-NA is Date (unlist would coerce to numeric)
+  # logicalNA_POSIXct: We add tzone = ""
+  # POSIXct_POSIXct (unlist would coerce to numeric)
+  pairs_result <- list("factor_character" = c("a", "b"),
```
Done 698e75023ac8b6363eb5e9238326df0050cd519c

zeehio commented on tests/testthat/test-combine.R:
```diff
@@ -16,3 +16,219 @@ test_that("combine works with input that used to fail (#1780)", {
+  expect_equal(works3, expected_result)
+})
+
+test_that("combine is strict combining/promoting types", {
```
Sorry, I do not know how to do that and I could not find any example using testthat to guide me.

However, when I run the `test-combine.R` script the `expect_that` functions that fail print the `info` argument that contains `Pair: (item1, item2)`. This is printed when running `R CMD check`:

```
  1. Failure: combine is strict combining/promoting types (@test-combine.R#214) --
  combine(items[c(pairs$Var1[i], pairs$Var2[i])]) did not throw an error.
  Pair: ( integer , character )
```
So by reading this I know that coercing an integer and a character did not throw an error as expected (this is a dummy example, it actually gives an error)

I imagine that you would rather have `combine(items[c("integer", "character")]) did not throw an error.` as an error message, but I do not know how to do that substitution. :sweat_smile:

zeehio replied:
I know how to do that now. Fix is on the way

hadley commented on inst/include/dplyr/Collecter.h:
```diff
@@ -363,13 +427,25 @@ namespace dplyr {
+        return new Collecter_Impl<INTSXP>(n);
+      else {
+        SEXP classes = Rf_getAttrib(model, R_ClassSymbol);
+        stop("Can't collect elements of class %s", CHAR(STRING_ELT(classes, 0)));
```
I think we should make this a warning for now

zeehio commented:
This is the summary of all the 11x11 = 121 possible coercions I have tested. I have considered all pairs of the following types:

- logicalvalue (logical with a non-NA value)
- logicalNA (NA)
- integer
- double
- complex
- factor
- character
- POSIXct
- Date
- num_with_class (custom numeric class, think of difftime)
- int_with_class (custom integer class)

The behaviour implemented in this PR allows to combine 35 out of 121 possible pairs.

# Allow combine with NA

Custom classes may lose attributes or have wrong attributes if a NA is present, therefore they give a warning.


|Var1           |Var2           |can_combine |warning |result                 |result_class    |
|:--------------|:--------------|:-----------|:-------|:----------------------|:---------------|
|logicalvalue   |logicalNA      |TRUE        |FALSE   |TRUE, NA               |logical         |
|logicalNA      |logicalvalue   |TRUE        |FALSE   |NA, TRUE               |logical         |
|logicalNA      |logicalNA      |TRUE        |FALSE   |NA, NA                 |logical         |
|logicalNA      |integer        |TRUE        |FALSE   |NA, 4                  |integer         |
|logicalNA      |factor         |TRUE        |FALSE   |NA, a                  |factor          |
|logicalNA      |double         |TRUE        |FALSE   |NA, 4.5                |numeric         |
|logicalNA      |character      |TRUE        |FALSE   |NA, b                  |character       |
|logicalNA      |POSIXct        |TRUE        |FALSE   |NA, 2010-01-01         |POSIXct, POSIXt |
|logicalNA      |Date           |TRUE        |FALSE   |NA, 2016-01-01         |Date            |
|logicalNA      |complex        |TRUE        |FALSE   |NA, 1+2i               |complex         |
|logicalNA      |int_with_class |TRUE        |**TRUE**   |NA, 4                  |int_with_class  |
|logicalNA      |num_with_class |TRUE        |**TRUE**   |NA, 4.5                |num_with_class  |
|integer        |logicalNA      |TRUE        |FALSE   |4, NA                  |integer         |
|factor         |logicalNA      |TRUE        |FALSE   |a, NA                  |factor          |
|double         |logicalNA      |TRUE        |FALSE   |4.5, NA                |numeric         |
|character      |logicalNA      |TRUE        |FALSE   |b, NA                  |character       |
|POSIXct        |logicalNA      |TRUE        |FALSE   |2010-01-01, NA         |POSIXct, POSIXt |
|Date           |logicalNA      |TRUE        |FALSE   |2016-01-01, NA         |Date            |
|complex        |logicalNA      |TRUE        |FALSE   |1+2i, NA               |complex         |
|int_with_class |logicalNA      |TRUE        |**TRUE**   |4, NA                  |int_with_class  |
|num_with_class |logicalNA      |TRUE        |**TRUE**   |4.5, NA                |num_with_class  |

# Allow combine with the same type

Custom classes may lose attributes and therefore give a warning.

|Var1           |Var2           |can_combine |warning |result                 |result_class    |
|:--------------|:--------------|:-----------|:-------|:----------------------|:---------------|
|logicalvalue   |logicalvalue   |TRUE        |FALSE   |TRUE, TRUE             |logical         |
|logicalNA      |logicalNA      |TRUE        |FALSE   |NA, NA                 |logical         |
|integer        |integer        |TRUE        |FALSE   |4, 4                   |integer         |
|factor         |factor         |TRUE        |FALSE   |a, a                   |factor          |
|double         |double         |TRUE        |FALSE   |4.5, 4.5               |numeric         |
|character      |character      |TRUE        |FALSE   |b, b                   |character       |
|POSIXct        |POSIXct        |TRUE        |FALSE   |2010-01-01, 2010-01-01 |POSIXct, POSIXt |
|Date           |Date           |TRUE        |FALSE   |2016-01-01, 2016-01-01 |Date            |
|complex        |complex        |TRUE        |FALSE   |1+2i, 1+2i             |complex         |
|int_with_class |int_with_class |TRUE        |**TRUE**    |4, 4                   |int_with_class  |
|num_with_class |num_with_class |TRUE        |**TRUE**    |4.5, 4.5               |num_with_class  |

# Allowed coercions

|Var1           |Var2           |can_combine |warning |result                 |result_class    |
|:--------------|:--------------|:-----------|:-------|:----------------------|:---------------|
|factor         |character      |TRUE        |**TRUE**   |a, b                   |character       |
|character      |factor         |TRUE        |FALSE   |b, a                   |character       |
|integer        |double         |TRUE        |FALSE   |4, 4.5                 |numeric         |
|double         |integer        |TRUE        |FALSE   |4.5, 4                 |numeric         |

# Not allowed combinations

|Var1           |Var2           |can_combine |warning |result                 |result_class    |
|:--------------|:--------------|:-----------|:-------|:----------------------|:---------------|
|logicalvalue   |integer        |FALSE       |FALSE   |                       |                |
|logicalvalue   |factor         |FALSE       |FALSE   |                       |                |
|logicalvalue   |double         |FALSE       |FALSE   |                       |                |
|logicalvalue   |character      |FALSE       |FALSE   |                       |                |
|logicalvalue   |POSIXct        |FALSE       |FALSE   |                       |                |
|logicalvalue   |Date           |FALSE       |FALSE   |                       |                |
|logicalvalue   |complex        |FALSE       |FALSE   |                       |                |
|logicalvalue   |int_with_class |FALSE       |FALSE   |                       |                |
|logicalvalue   |num_with_class |FALSE       |FALSE   |                       |                |
|integer        |logicalvalue   |FALSE       |FALSE   |                       |                |
|integer        |factor         |FALSE       |FALSE   |                       |                |
|integer        |character      |FALSE       |FALSE   |                       |                |
|integer        |POSIXct        |FALSE       |FALSE   |                       |                |
|integer        |Date           |FALSE       |FALSE   |                       |                |
|integer        |complex        |FALSE       |FALSE   |                       |                |
|integer        |int_with_class |FALSE       |FALSE   |                       |                |
|integer        |num_with_class |FALSE       |FALSE   |                       |                |
|factor         |logicalvalue   |FALSE       |FALSE   |                       |                |
|factor         |integer        |FALSE       |FALSE   |                       |                |
|factor         |double         |FALSE       |FALSE   |                       |                |
|factor         |POSIXct        |FALSE       |FALSE   |                       |                |
|factor         |Date           |FALSE       |FALSE   |                       |                |
|factor         |complex        |FALSE       |FALSE   |                       |                |
|factor         |int_with_class |FALSE       |FALSE   |                       |                |
|factor         |num_with_class |FALSE       |FALSE   |                       |                |
|double         |logicalvalue   |FALSE       |FALSE   |                       |                |
|double         |factor         |FALSE       |FALSE   |                       |                |
|double         |character      |FALSE       |FALSE   |                       |                |
|double         |POSIXct        |FALSE       |FALSE   |                       |                |
|double         |Date           |FALSE       |FALSE   |                       |                |
|double         |complex        |FALSE       |FALSE   |                       |                |
|double         |int_with_class |FALSE       |FALSE   |                       |                |
|double         |num_with_class |FALSE       |FALSE   |                       |                |
|character      |logicalvalue   |FALSE       |FALSE   |                       |                |
|character      |integer        |FALSE       |FALSE   |                       |                |
|character      |double         |FALSE       |FALSE   |                       |                |
|character      |POSIXct        |FALSE       |FALSE   |                       |                |
|character      |Date           |FALSE       |FALSE   |                       |                |
|character      |complex        |FALSE       |FALSE   |                       |                |
|character      |int_with_class |FALSE       |FALSE   |                       |                |
|character      |num_with_class |FALSE       |FALSE   |                       |                |
|POSIXct        |logicalvalue   |FALSE       |FALSE   |                       |                |
|POSIXct        |integer        |FALSE       |FALSE   |                       |                |
|POSIXct        |factor         |FALSE       |FALSE   |                       |                |
|POSIXct        |double         |FALSE       |FALSE   |                       |                |
|POSIXct        |character      |FALSE       |FALSE   |                       |                |
|POSIXct        |Date           |FALSE       |FALSE   |                       |                |
|POSIXct        |complex        |FALSE       |FALSE   |                       |                |
|POSIXct        |int_with_class |FALSE       |FALSE   |                       |                |
|POSIXct        |num_with_class |FALSE       |FALSE   |                       |                |
|Date           |logicalvalue   |FALSE       |FALSE   |                       |                |
|Date           |integer        |FALSE       |FALSE   |                       |                |
|Date           |factor         |FALSE       |FALSE   |                       |                |
|Date           |double         |FALSE       |FALSE   |                       |                |
|Date           |character      |FALSE       |FALSE   |                       |                |
|Date           |POSIXct        |FALSE       |FALSE   |                       |                |
|Date           |complex        |FALSE       |FALSE   |                       |                |
|Date           |int_with_class |FALSE       |FALSE   |                       |                |
|Date           |num_with_class |FALSE       |FALSE   |                       |                |
|complex        |logicalvalue   |FALSE       |FALSE   |                       |                |
|complex        |integer        |FALSE       |FALSE   |                       |                |
|complex        |factor         |FALSE       |FALSE   |                       |                |
|complex        |double         |FALSE       |FALSE   |                       |                |
|complex        |character      |FALSE       |FALSE   |                       |                |
|complex        |POSIXct        |FALSE       |FALSE   |                       |                |
|complex        |Date           |FALSE       |FALSE   |                       |                |
|complex        |int_with_class |FALSE       |FALSE   |                       |                |
|complex        |num_with_class |FALSE       |FALSE   |                       |                |
|int_with_class |logicalvalue   |FALSE       |FALSE   |                       |                |
|int_with_class |integer        |FALSE       |FALSE   |                       |                |
|int_with_class |factor         |FALSE       |FALSE   |                       |                |
|int_with_class |double         |FALSE       |FALSE   |                       |                |
|int_with_class |character      |FALSE       |FALSE   |                       |                |
|int_with_class |POSIXct        |FALSE       |FALSE   |                       |                |
|int_with_class |Date           |FALSE       |FALSE   |                       |                |
|int_with_class |complex        |FALSE       |FALSE   |                       |                |
|int_with_class |num_with_class |FALSE       |FALSE   |                       |                |
|num_with_class |logicalvalue   |FALSE       |FALSE   |                       |                |
|num_with_class |integer        |FALSE       |FALSE   |                       |                |
|num_with_class |factor         |FALSE       |FALSE   |                       |                |
|num_with_class |double         |FALSE       |FALSE   |                       |                |
|num_with_class |character      |FALSE       |FALSE   |                       |                |
|num_with_class |POSIXct        |FALSE       |FALSE   |                       |                |
|num_with_class |Date           |FALSE       |FALSE   |                       |                |
|num_with_class |complex        |FALSE       |FALSE   |                       |                |
|num_with_class |int_with_class |FALSE       |FALSE   |                       |                |

I have rebased and squashed the commits into 4 manageable commits. I will appreciate your comments.

Oh, and in case a regression appeared in the future, the error message would explicit (here is an example of what would be printed):
```
1. Failure: Coercion from factor to character (@test-combine.R#282) ----------
combine(items[c("factor", "character")]) not equal to c("a", "b").
```

zeehio commented:
This has been left here ready for review, but no hurries (end of year is usually busy, so happy new year to all of you)

krlmlr commented on inst/include/dplyr/Collecter.h:
```diff
@@ -347,11 +388,34 @@ namespace dplyr {
+          data[ index[i] ] = levels_map.find(x)->second;
+        }
+      }
+    }
```
And here, too.

krlmlr commented on inst/include/dplyr/Collecter.h:
```diff
@@ -399,18 +475,35 @@ namespace dplyr {
         return new Collecter_Impl<STRSXP>(n);
+      if (has_classes(model)) {
+        SEXP classes = Rf_getAttrib(model, R_ClassSymbol);
+        Rf_warning("Coercing class %s, with possible loss of information",
```
Coercing into what?

krlmlr commented on inst/include/dplyr/Collecter.h:
```diff
@@ -399,18 +475,35 @@ namespace dplyr {
         return new TypedCollecter<REALSXP>(n, get_date_classes());
+      if (has_classes(model)) {
+        SEXP classes = Rf_getAttrib(model, R_ClassSymbol);
+        Rf_warning("Coercing class %s, with possible loss of information",
```
... into what?

krlmlr commented:
Oh, don't worry about the merge conflict, I can resolve it.

zeehio commented on tests/testthat/test-combine.R:
```diff
@@ -16,3 +16,276 @@ test_that("combine works with input that used to fail (#1780)", {
+                 label = paste0("combine(items[c(\"", var1, "\", \"", var2, "\")])"),
+                 expected.label = deparse(result))
+  } else {
+    expect_error(suppressWarnings(combine(item_pair)),
```
`expect_error(expect_warning(...))` will not work as expected, because if a bug causes `...` to run without warnings, then `expect_warning` will raise an error that will be accepted and suppressed by `expect_error` and the test will succeed where it should have failed.

I am working on all the other comments

krlmlr replied:
How about `expect_warning(expect_error(...))` ?

``` r
library(testthat)

with_reporter("summary",
  test_that("error + warning", {
    expect_warning(expect_error({warning(); stop()}))
    expect_warning(expect_error({stop()}))
    expect_warning(expect_error({warning()}))
    expect_warning(expect_error({}))
  })
)
#> ...12.34
#> Failed --------------------------------------------------------------------
#> 1. Failure: error + warning (@<text>#6) -----------------------------------
#> expect_error(...) showed 0 warnings
#> 
#> 
#> 2. Failure: error + warning (@<text>#7) -----------------------------------
#> {
#>     ...
#> } did not throw an error.
#> 
#> 
#> 3. Failure: error + warning (@<text>#8) -----------------------------------
#> {
#>     ...
#> } did not throw an error.
#> 
#> 
#> 4. Failure: error + warning (@<text>#8) -----------------------------------
#> expect_error(...) showed 0 warnings
#> 
#> 
#> DONE ======================================================================
```

zeehio commented:
I have taken care of all your comments. The most noticeable change is that I decoupled `prepare_table_with_coercion_rules` from the `items` variable.

I am ready for the next round of comments if there still are any other left :+1:

krlmlr commented on tests/testthat/helper-combine.R:
```diff
@@ -0,0 +1,200 @@
+                int_with_class = structure(4L, class = "int_with_class"),
+                num_with_class = structure(4.5, class = "num_with_class"))
+
+  pairs <- prepare_table_with_coercion_rules(items)
```
Can you define items in prepare_table_with_coercion_rules() and call that function without arguments? This probably means that we need to add a new list column "item_pair" to the pairs data frame.

zeehio replied:
Done

krlmlr commented on tests/testthat/helper-combine.R:
```diff
@@ -0,0 +1,200 @@
+  can_be_combined <- function(item1, item2,
+                              class1, class2,
+                              all_na1, all_na2,
+                              known_to_dplyr1, known_to_dplyr2) {
```
I'm not a big fan of local functions. Do they access variables other than their arguments?

zeehio replied:
As you prefer, I have placed those local functions as conventional functions. :+1: 

Like you, I don't like functions that use variables out of their scope (variables other than their arguments). However I find it practical to have local functions when there is only one function that calls them (a local function, like a local variable).

As this is just a matter of taste, I have avoided using local functions as you requested :smiley:

zeehio commented:
I will finish on Monday evening. A romantic weekend in Rome awaits me.

Thanks for all the time and reviews!

krlmlr commented:
Thanks!
